lib/templates.php

```diff
@@ -75,9 +75,41 @@ function gutenberg_register_template_post_type() {
 	);
 
 	register_post_type( 'wp_template', $args );
+
+	$meta_args = array(
+		'object_subtype' => 'wp_template',
+		'type'           => 'string',
+		'description'    => 'The theme that provided the template, if any.',
+		'single'         => true,
+		'show_in_rest'   => true,
+	);
+
+	register_post_type( 'wp_template', $args );
+	register_meta( 'post', 'theme', $meta_args );
 }
 add_action( 'init', 'gutenberg_register_template_post_type' );
 
+/**
+ * Automatically set the theme meta for templates.
+ *
+ * @param array $post_id Template ID.
+ * @param array $post    Template Post.
+ * @param bool  $update  Is update.
+ */
+function gutenberg_set_template_post_theme( $post_id, $post, $update ) {
+	if ( 'wp_template' !== $post->post_type || $update || 'trash' === $post->post_status ) {
+		return;
+	}
+
+	$theme = get_post_meta( $post_id, 'theme', true );
+
+	if ( ! $theme ) {
+		update_post_meta( $post_id, 'theme', wp_get_theme()->get_stylesheet() );
+	}
+}
```

Comment:
Im now realizing this means that custom template parts created through wp-admin will be given a theme value corresponding to the active theme.  This will make it hard to differentiate TPs that were originally supplied by a theme vs. custom user created ones that should not be tied to a theme, which will in turn make it difficult to show a proper list of template parts in the site editor ( current theme + custom ). 🤔 

I wonder how we can get around that.

Reply:
When I wrote this, template parts were theme-specific on my mind even user created one but if they are not, maybe we can just show all the template parts from all themes (non auto-drafts) + the  auto-drafts of the current theme in the UI.

Reply:
This will be fixed by #27016, which has a new taxonomy term `wp_file_based` to indicate ones that originated from the theme. Ones that originated from the user won't have this taxonomy term.

Reply:
that taxonomy is weird, what if I have something that originates from the theme but then I completely edit it, at what time it stops to be something that originates from the theme and that becomes user based?

Reply:
I've replied about it in another PR, so let me just link that comment here: https://github.com/WordPress/gutenberg/pull/27016#issuecomment-732068742

The tl;dr is that `_wp_file_based` only indicate the origin of a template/part, which is useful for things like reverting it to the file version.
Once a template/part is not `auto-draft` anymore, it's safe to say it's "user based", regardless of the `_wp_file_based` tag.

(I'm totally open to change the tag name if y'all feel it's confusing!)

Reply:
> When I wrote this, template parts were theme-specific on my mind even user created one but if they are not, maybe we can just show all the template parts from all themes (non auto-drafts) + the auto-drafts of the current theme in the UI.

I don't think there has been much discussion on whether they should or shouldn't be tied to the theme.  However, in the past slug/theme combos were required to be unique.  When this restriction was taken off, custom template parts were created with a `''` theme name making them distinguishable from a theme.  Now that they do have the same theme value, there are other conflicts that can come into play such as unedited theme supplied templates potentially referencing the wrong template parts as described in the second half of my comment here -https://github.com/WordPress/gutenberg/issues/26950#issuecomment-733081422